### PR TITLE
Improve eigensolver user interface 

### PR DESCRIPTION
**Pull request by fdkong**
Added eigenvalue  executioner. Migrate solver options from `eigenproblem` to `eigenexecutioner `.

Add native `moose` parameters to control solver behaviors instead of `petsc` options.

### CODE REVIEW

File: framework/src/actions/CreateExecutionerAction.C [hunk 1]
```diff
@@ -38,8 +40,14 @@ CreateExecutionerAction::act()
   _moose_object_pars.set<FEProblemBase *>("_fe_problem_base") = _problem.get();
   _moose_object_pars.set<FEProblem *>("_fe_problem") =
       (std::dynamic_pointer_cast<FEProblem>(_problem)).get();
+  _moose_object_pars.set<EigenProblem *>("_eigen_problem") =
+      (std::dynamic_pointer_cast<EigenProblem>(_problem)).get();
```
Comment (permcody) [inline framework/src/actions/CreateExecutionerAction.C]: I know this is a copy of the line above but it's either unsafe or incorrect. You are dynamic casting which suggests that you recognize that it might not always succeed, yet you assume it does by immediately dereferencing the pointer (which may be nullptr). I'd suggest either checking the cast, or changing to a static cast to indicate that you know that it will always succeed. You should change the other similar lines too.

File: framework/src/actions/CreateExecutionerAction.C [hunk 2]
```diff
@@ -38,8 +40,14 @@ CreateExecutionerAction::act()
...
   std::shared_ptr<Executioner> executioner =
       _factory.create<Executioner>(_type, "Executioner", _moose_object_pars);
 
+  if (std::dynamic_pointer_cast<EigenProblem>(_problem).get() &&
+      !(std::dynamic_pointer_cast<EigenExecutioner>(executioner).get()))
+    mooseError("EigenProblem requires an EigenExecutioner or its derived subclass");
```
Comment (permcody) [inline framework/src/actions/CreateExecutionerAction.C]: Oh, I see, here's the check. We'll no need to perform these casts twice and in the wrong order. Save your cast results.

File: framework/src/actions/CreateProblemAction.C [hunk 3]
```diff
@@ -68,6 +69,28 @@ CreateProblemAction::act()
...
+      if (_app.useEigenExecutioner() && _type == "FEProblem")
+      {
+        _type = "EigenProblem";
+        parameters().set<std::string>("type") = "EigenProblem";
+        _moose_object_pars =
+            (!parameters().have_parameter<bool>("skip_param_construction") ||
+                     (parameters().have_parameter<bool>("skip_param_construction") &&
+                      !parameters().get<bool>("skip_param_construction"))
+                 ? _factory.getValidParams(_type)
+                 : validParams<MooseObject>());
```
Comment (permcody) [inline framework/src/actions/CreateProblemAction.C]: This is such a nasty line of code that it really shouldn't be repeated (copied). At the very least, it should be factored into a helper method in the original class and called here.

File: framework/src/actions/CreateProblemAction.C [hunk 4]
```diff
@@ -68,6 +69,28 @@ CreateProblemAction::act()
...
+        // if there is a [Problem] block, the parameters will be retrieved from parser
+        if (parameters().isParamValid("parser_syntax"))
+        {
+          _app.parser().extractParams(parameters().get<std::string>("parser_syntax"),
```
Comment (permcody) [inline framework/src/actions/CreateProblemAction.C]: Note that I dislike invoking the parser directly (as we've already discussed, I'll take a look at this and see if I can come up with a better way).

File: framework/src/actions/CreateProblemAction.C [hunk 5]
```diff
@@ -81,6 +104,13 @@ CreateProblemAction::act()
           _factory.create<FEProblemBase>(_type, getParam<std::string>("name"), _moose_object_pars);
       if (!_problem.get())
         mooseError("Problem has to be of a FEProblemBase type");
+
+      // if users provide a problem type, the type has to be an EigenProblem or its derived subclass
+      // when uing an eigen executioner
+      if (_app.useEigenExecutioner() && _type != "EigenProblem" &&
+          !(std::dynamic_pointer_cast<EigenProblem>(_problem).get()))
```
Comment (permcody) [inline framework/src/actions/CreateProblemAction.C]: Note that you don't have to use `.get()` to test for `nullptr`. You can test the smart pointer directly. This just cleans up the code. You do this elsewhere in this PR as well.

File: framework/src/base/EigenProblem.C [hunk 6]
```diff
@@ -30,40 +33,26 @@ InputParameters
...
   return params;
 }
 
 EigenProblem::EigenProblem(const InputParameters & parameters)
   : FEProblemBase(parameters),
-    _n_eigen_pairs_required(getParam<unsigned int>("n_eigen_pairs")),
+    _n_eigen_pairs_required(1),
     _generalized_eigenvalue_problem(false),
-    _nl_eigen(new NonlinearEigenSystem(*this, "eigen0"))
+    _nl_eigen(new NonlinearEigenSystem(*this, "eigen0")),
```
Comment (permcody) [inline framework/src/base/EigenProblem.C]: Can we cleanup these raw calls to `new`? Prefer `make_shared` instead, it's not just style, it's safer.

File: framework/src/base/EigenProblem.C [hunk 7]
```diff
@@ -30,40 +33,26 @@ InputParameters
...
+    _is_residual_initialed(false)
 {
 #if LIBMESH_HAVE_SLEPC
   _nl = _nl_eigen;
   _aux = new AuxiliarySystem(*this, "aux0");
```
Comment (permcody) [inline framework/src/base/EigenProblem.C]: I know this isn't in your PR, but if you cleanup the `new` call above, you might as well clean this one up as well.

File: framework/include/base/EigenProblem.h [hunk 8]
```diff
@@ -41,15 +41,30 @@ class EigenProblem : public FEProblemBase
...
   virtual unsigned int getNEigenPairsRequired() { return _n_eigen_pairs_required; }
+  virtual void setNEigenPairsRequired(unsigned int n_eigen_pairs)
+  {
+    _n_eigen_pairs_required = n_eigen_pairs;
+  }
   virtual bool isGeneralizedEigenvalueProblem() { return _generalized_eigenvalue_problem; }
   virtual bool isNonlinearEigenvalueSolver();
 
+  bool residualInitialed() { return _is_residual_initialed; }
+  void setResidualInitialed(bool is_residual_initialed)
```
Comment (permcody) [inline framework/include/base/EigenProblem.h]: Can you document this method? I'm not sure what it's for, it seems odd to me that you can initialize what I would assume is internal state of the object from a method call. Why is this needed?

File: framework/include/executioners/EigenExecutioner.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#ifndef EIGENEXECUTIONER_H
```
Comment (permcody) [inline framework/include/executioners/EigenExecutioner.h]: What's the plan for this new Executioner? We have a few others already (nonlinearEigen and powermethod?) Is this going to be a new base class, will it replace the others?

File: framework/src/utils/SlepcSupport.C [hunk 10]
```diff
@@ -356,9 +433,17 @@ moosePetscSNESFormFunction(SNES /*snes*/, Vec x, Vec r, void * ctx, Moose::Kerne
...
+  if (type == Moose::KT_EIGEN && eigen_problem->residualInitialed())
+  {
+    eigen_problem->computeResidualTypeBx(*sys.current_local_solution.get(), R, type);
+    eigen_problem->setResidualInitialed(false);
+  }
+  else
+  {
+    eigen_problem->computeResidualType(*sys.current_local_solution.get(), R, type);
+    if (type == Moose::KT_NONEIGEN)
+      eigen_problem->setResidualInitialed(true);
```
Comment (permcody) [inline framework/src/utils/SlepcSupport.C]: Ah, here's where you are using that odd method. You are passing in type, can you not set the state internally when you call computeResidualType*? Maybe not, just asking

**Review by permcody — CHANGES_REQUESTED** [review]

File: framework/include/executioners/EigenExecutioner.h [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+class EigenProblem;
+
+template <typename T>
+InputParameters validParams();
+
+template <>
+InputParameters validParams<EigenExecutioner>();
+
+
+class EigenExecutioner : public Steady
```
Comment (YaqiWang) [inline framework/include/executioners/EigenExecutioner.h]: Before reviewing, I do not quite like the name. We will have something like
```
[Executioner]
  type = EigenExecutioner
  ...
[]
```
Executioner word appeared twice. Maybe it is better just use `Eigenvalue`?

File: framework/src/actions/CreateExecutionerAction.C (hunk 1 above)
Comment (fdkong) [inline framework/src/actions/CreateExecutionerAction.C]: If succeed, I know this is an `eigenproblem`, otherwise it may be a regular `FEProblem`. I need this result to check if the `executioner` is consistent with `problem`.

I need `EigenExecutioner` work with `EigenProblem` only. 

Any other better way to get this goal?

File: framework/src/actions/CreateProblemAction.C (hunk 4 above)
Comment (fdkong) [inline framework/src/actions/CreateProblemAction.C]: The code using ` meta action` is even more nasty.

File: framework/src/actions/CreateProblemAction.C (hunk 3 above)
Comment (fdkong) [inline framework/src/actions/CreateProblemAction.C]: I just simply copied what we already have at the first  place.

File: framework/src/actions/CreateProblemAction.C (hunk 5 above)
Comment (fdkong) [inline framework/src/actions/CreateProblemAction.C]: OK

File: framework/src/base/EigenProblem.C (hunk 6 above)
Comment (fdkong) [inline framework/src/base/EigenProblem.C]: I did not touch this line in this PR. It will be a big refactor.

File: framework/include/executioners/EigenExecutioner.h (hunk 9 above)
Comment (fdkong) [inline framework/include/executioners/EigenExecutioner.h]: Will replace all others.

File: framework/src/utils/SlepcSupport.C (hunk 10 above)
Comment (fdkong) [inline framework/src/utils/SlepcSupport.C]: The method is not odd at all. The` eigenvalue  solver ` is different from the normal `nonlinear` solver. We need to  evaluate the residual twice, one for the left, and the other for the right.  If the left residual is called, we are better to skip a lot of extra computations such as `multiapp transfers`, `UO` and so on. 

The idea comes from @YaqiWang

File: framework/include/executioners/EigenExecutioner.h (hunk 11 above)
Comment (fdkong) [inline framework/include/executioners/EigenExecutioner.h]: Eigenvalue has been used in `post processor `.

File: framework/src/utils/SlepcSupport.C (hunk 10 above)
Comment (permcody) [inline framework/src/utils/SlepcSupport.C]: What's odd about it is the fact that it's an internal state variable that's wide open to be set by anyone. I don't understand why you can't set the internal state from within the object when you compute residuals based on all of the parameters that you pass in when computing those residuals. To me, this variable should be exposed as a setter from an encapsulation point of view.

File: framework/src/base/EigenProblem.C (hunk 6 above)
Comment (fdkong) [inline framework/src/base/EigenProblem.C]: Because `FEProblemBase` also owns this pointer, we may not use a unique pointer.

Comment (permcody) [inline framework/src/base/EigenProblem.C]: Own or use? The difference is important. Own -> smart parter, use or share -> raw pointer. I argue that it's the latter. The pointer itself lives in this class, it's created when the problem is created and deleted when the problem is destroyed. I don't think FEProblemBase can ever own this pointer.

File: framework/src/utils/SlepcSupport.C [hunk 12]
```diff
@@ -96,10 +96,53 @@ getSlepcEigenProblemValidParams()
   params.addParam<unsigned int>("n_eigen_pairs", 1, "The number of eigen pairs");
   params.addParam<unsigned int>("n_basis_vectors", 3, "The dimension of eigen subspaces");
 
+  params.addParam<Real>("e_tol", 1.0e-4, "Relative Tolerance for Eigen Solver");
+  params.addParam<unsigned int>("e_max_its", 10000, "Max Linear Iterations for Eigen Solver");
```
Comment (YaqiWang) [inline framework/src/utils/SlepcSupport.C]: Oh, I see, you used `l_max_its` to control power iteration. Then I suggest to changes this to `l_max_power_its` for power iteration and use `l_max_its` for Newton.

File: framework/src/utils/SlepcSupport.C [hunk 13]
```diff
@@ -96,10 +96,53 @@ getSlepcEigenProblemValidParams()
...
+
+    // linear solver
+    Moose::PetscSupport::setSinglePetscOption("-eps_power_ksp_max_it",
+                                              stringify(params.get<unsigned int>("l_max_its")));
+
+    Moose::PetscSupport::setSinglePetscOption("-eps_power_ksp_rtol",
+                                              stringify(params.get<Real>("l_tol")));
+
+    Moose::PetscSupport::setSinglePetscOption("-eps_power_ksp_atol",
+                                              stringify(params.get<Real>("l_abs_step_tol")));
```
Comment (YaqiWang) [inline framework/src/utils/SlepcSupport.C]: Step tol is different from abs tol, are you sure here?

Comment (YaqiWang) [inline framework/src/utils/SlepcSupport.C]: I suggest to keep these linear iteration control parameters for Newton and create a new set of parameters for controlling power iteration if really necessary. We can do some off-line discussions on this.

File: test/tests/problems/eigen_problem/ipm.i [hunk 14]
```diff
@@ -79,6 +79,7 @@
 
 [Executioner]
   type = EigenExecutioner
+  eigen_solve_type = KRYLOVSCHUR
```
Comment (YaqiWang) [inline test/tests/problems/eigen_problem/ipm.i]: Since eigen executioner is separated from Steady, we can just use `solve_type` here.

File: framework/src/actions/CreateProblemAction.C (hunk 4 above)
Comment (YaqiWang) [inline framework/src/actions/CreateProblemAction.C]: We three need to discuss this. I think the meta action way will be much cleaner and more extendable.

File: framework/src/utils/SlepcSupport.C [hunk 15]
```diff
@@ -105,6 +109,23 @@ storeSlepcEigenProblemOptions(EigenProblem & eigen_problem, const InputParameter
...
+  unsigned int n_eigen_pairs = params.get<unsigned int>("n_eigen_pairs");
+  unsigned int n_basis_vectors = params.get<unsigned int>("n_basis_vectors");
+
+  eigen_problem.setNEigenPairsRequired(n_eigen_pairs);
+
+  eigen_problem.es().parameters.set<unsigned int>("eigenpairs") = n_eigen_pairs;
+
+  // If the subspace dimension is too small, we increase it automatically
+  if (2 * n_eigen_pairs > n_basis_vectors)
+    n_basis_vectors = 2 * n_eigen_pairs;
```
Comment (YaqiWang) [inline framework/src/utils/SlepcSupport.C]: Can we issue a warning here saying that the user provided parameter has been overwritten by the code due to bla bla?

File: framework/src/utils/SlepcSupport.C [hunk 16]
```diff
@@ -96,19 +96,21 @@ getSlepcEigenProblemValidParams()
   params.addParam<unsigned int>("n_eigen_pairs", 1, "The number of eigen pairs");
   params.addParam<unsigned int>("n_basis_vectors", 3, "The dimension of eigen subspaces");
 
-  params.addParam<Real>("e_tol", 1.0e-4, "Relative Tolerance for Eigen Solver");
-  params.addParam<unsigned int>("e_max_its", 10000, "Max Linear Iterations for Eigen Solver");
+  params.addParam<Real>("eigen_tol", 1.0e-4, "Relative Tolerance for Eigen Solver");
+  params.addParam<unsigned int>("eigen_max_its", 10000, "Max Linear Iterations for Eigen Solver");
```
Comment (YaqiWang) [inline framework/src/utils/SlepcSupport.C]: Ah, this param is changed from `l_max_its` to `eigen_max_its`. Please ignore my comments in your previous related commits.

File: framework/src/utils/SlepcSupport.C [hunk 17]
```diff
@@ -142,18 +144,33 @@ setSlepcEigenSolverTolerances(EigenProblem & eigen_problem, const InputParameter
     Moose::PetscSupport::setSinglePetscOption("-eps_power_ksp_atol",
                                               stringify(params.get<Real>("l_abs_step_tol")));
   }
+  else
+  { // linear eigenvalue problem
+    // linear solver
+    Moose::PetscSupport::setSinglePetscOption("-st_ksp_max_it",
```
Comment (YaqiWang) [inline framework/src/utils/SlepcSupport.C]: do we have a test on this?

**Review by YaqiWang — COMMENTED** [review]
@fdkong This PR represents a significant improvements on both the user interface and the code design. On the design side, I think we should use meta action, `Eigenvalue` name and a function for Ax-Bx/lambda. The user interface is great. I just want you to have a documentation on ALL the parameters under Executioner like what we had in Rattlesnake user manual. You will probably find necessary changes in the code during writing.

File: framework/src/utils/SlepcSupport.C (hunk 15 above)
Comment (fdkong) [inline framework/src/utils/SlepcSupport.C]: Maybe a good idea

File: framework/src/utils/SlepcSupport.C (hunk 13 above)
Comment (fdkong) [inline framework/src/utils/SlepcSupport.C]: Typo

File: framework/src/actions/DetermineSystemType.C [hunk 18]
```diff
@@ -37,4 +37,8 @@ DetermineSystemType::act()
    */
   if (_moose_object_pars.isParamValid("_eigen") && _moose_object_pars.get<bool>("_eigen"))
     _app.useNonlinear() = false;
+
+  if (_moose_object_pars.isParamValid("_use_eigen_executioner") &&
+      _moose_object_pars.get<bool>("_use_eigen_executioner"))
+    _app.useEigenExecutioner() = true;
```
Comment (fdkong) [inline framework/src/actions/DetermineSystemType.C]: I guess the meta action also needs to call parser to retrieve input parameters.

**Comment by fdkong** [comment]
@YaqiWang  Please work on the updated code. Since you looks like working the outdated code, and then I do not know how to response to you

**Comment by YaqiWang** [comment]
@fdkong I did. I realized few of my comments are on the commits which later commits changes the modifications of.

File: framework/src/utils/SlepcSupport.C (hunk 13 above)
Comment (fdkong) [inline framework/src/utils/SlepcSupport.C]: I checked again. This should be fine.

File: framework/src/utils/SlepcSupport.C (hunk 12 above)
Comment (fdkong) [inline framework/src/utils/SlepcSupport.C]: `e_max_its` power iteration, `nl_max_its` for nonlinear solver, `l_max_its` for linear solver

**Comment by fdkong** [comment]
@permcody, PR is complete, and I do not have anything to add any more.

File: framework/src/actions/DetermineSystemType.C [hunk 19]
```diff
@@ -37,4 +39,27 @@ DetermineSystemType::act()
...
+  {
+    _action_params.set<std::string>("type") = "FEProblem";
+  }
+
+  // if we have Problem block in input file, "CreateProblemAction" will be handled by parser
+  if (_awh.hasActions("create_problem"))
+    return;
+
+  // Create "CreateProblemAction"
+  std::shared_ptr<MooseObjectAction> action =
```
Comment (permcody) [inline framework/src/actions/DetermineSystemType.C]: No point to casting. You don't use the casted value. Just create and add instead.

File: framework/src/base/EigenProblem.C [hunk 20]
```diff
@@ -30,40 +33,26 @@ InputParameters
...
-#if LIBMESH_HAVE_SLEPC
-  params += Moose::SlepcSupport::getSlepcEigenProblemValidParams();
-#endif
   return params;
 }
 
 EigenProblem::EigenProblem(const InputParameters & parameters)
   : FEProblemBase(parameters),
-    _n_eigen_pairs_required(getParam<unsigned int>("n_eigen_pairs")),
+    _n_eigen_pairs_required(1),
```
Comment (permcody) [inline framework/src/base/EigenProblem.C]: Magic number? Maybe use a constant?

File: framework/src/base/EigenProblem.C [hunk 21]
```diff
@@ -30,40 +33,26 @@ InputParameters
...
     _generalized_eigenvalue_problem(false),
-    _nl_eigen(new NonlinearEigenSystem(*this, "eigen0"))
+    _nl_eigen(libmesh_make_unique<NonlinearEigenSystem>(*this, "eigen0")),
+    _is_residual_initialed(false)
 {
 #if LIBMESH_HAVE_SLEPC
-  _nl = _nl_eigen;
+  _nl = _nl_eigen.get();
```
Comment (permcody) [inline framework/src/base/EigenProblem.C]: Hmm, several issues here. I happen to know that we delete the _nl pointer in FEProblemBase. This is probably a disaster waiting to happen. Also, if _nl and _nl_eigen are equivalent, why are you using _nl_eigen at all? This needs to be refactored and designed properly.

File: framework/src/actions/CreateExecutionerAction.C [hunk 22]
```diff
@@ -36,10 +38,24 @@ void
...
+  if (eigen_problem)
+    _moose_object_pars.set<EigenProblem *>("_eigen_problem") = eigen_problem.get();
+
   std::shared_ptr<Executioner> executioner =
       _factory.create<Executioner>(_type, "Executioner", _moose_object_pars);
 
+  std::shared_ptr<Eigenvalue> eigen_executioner =
+      std::dynamic_pointer_cast<Eigenvalue>(executioner);
+
+  if ((eigen_problem && !eigen_executioner) || (!eigen_problem && eigen_executioner))
```
Comment (permcody) [inline framework/src/actions/CreateExecutionerAction.C]: You can do XOR with Boolean values: `eigen_problem != eigen_executioner`

File: framework/src/base/EigenProblem.C (hunk 20 above)
Comment (permcody) [inline framework/src/base/EigenProblem.C]: Looks like you moved these parameters to the SLEPC support class?

File: framework/src/base/EigenProblem.C [hunk 23]
```diff
@@ -171,3 +160,34 @@ EigenProblem::isNonlinearEigenvalueSolver()
          solverParams()._eigen_solve_type == Moose::EST_MONOLITH_NEWTON ||
          solverParams()._eigen_solve_type == Moose::EST_MF_MONOLITH_NEWTON;
 }
+
+void
+EigenProblem::computeResidualTypeBx(const NumericVector<Number> & soln,
+                                    NumericVector<Number> & residual,
+                                    Moose::KernelType type)
+{
+  _nl->setSolution(soln);
```
Comment (permcody) [inline framework/src/base/EigenProblem.C]: No you are using `_nl`, not `_nl_eigen`. I think you can delete the latter.

File: framework/src/utils/SlepcSupport.C [hunk 24]
```diff
@@ -91,20 +95,108 @@ getSlepcEigenProblemValidParams()
...
+  // ncv >= nev must hold and ncv >= 2*nev is recommended
+  unsigned int n_eigen_pairs = params.get<unsigned int>("n_eigen_pairs");
+  unsigned int n_basis_vectors = params.get<unsigned int>("n_basis_vectors");
+
+  eigen_problem.setNEigenPairsRequired(n_eigen_pairs);
+
+  eigen_problem.es().parameters.set<unsigned int>("eigenpairs") = n_eigen_pairs;
+
+  // If the subspace dimension is too small, we increase it automatically
+  if (2 * n_eigen_pairs > n_basis_vectors)
```
Comment (permcody) [inline framework/src/utils/SlepcSupport.C]: Magic number: consider a constant here.

File: framework/src/utils/SlepcSupport.C [hunk 25]
```diff
@@ -115,13 +207,7 @@ storeSlepcOptions(FEProblemBase & fe_problem, const InputParameters & params)
...
   if (params.isParamValid("solve_type"))
   {
-    mooseError("Can not use \"solve_type\" for eigenvalue problems, please use "
-               "\"eigen_solve_type\" instead \n");
-  }
-
-  if (params.isParamValid("eigen_solve_type"))
-  {
-    const std::string & eigen_solve_type = params.get<MooseEnum>("eigen_solve_type");
+    const std::string & eigen_solve_type = params.get<MooseEnum>("solve_type");
```
Comment (permcody) [inline framework/src/utils/SlepcSupport.C]: Are you sure you want to retrieve this as a string? It's stored as a MooseEnum type which is better.

File: framework/src/utils/SlepcSupport.C [hunk 26]
```diff
@@ -115,13 +207,7 @@ storeSlepcOptions(FEProblemBase & fe_problem, const InputParameters & params)
...
     fe_problem.solverParams()._eigen_solve_type =
         Moose::stringToEnum<Moose::EigenSolveType>(eigen_solve_type);
```
Comment (permcody) [inline framework/src/utils/SlepcSupport.C]: So you just turn it back into an enum here. I'd avoid the first conversion. You can always use `auto`!

**Review by permcody — CHANGES_REQUESTED** [review]

File: framework/include/actions/DetermineSystemType.h [hunk 27]
```diff
@@ -28,6 +28,9 @@ class DetermineSystemType : public MooseObjectAction
   DetermineSystemType(InputParameters parameters);
 
   virtual void act() override;
+
+private:
+  InputParameters _action_params;
```
Comment (fdkong) [inline framework/include/actions/DetermineSystemType.h]: `_pars` is for `DetermineSystemType`,  and `_action_params` for `CreateProblemAction`

Comment (fdkong) [inline framework/include/actions/DetermineSystemType.h]: `_moose_object_pars` for `Executioner`. I am missing something here?

File: framework/src/base/EigenProblem.C (hunk 20 above)
Comment (fdkong) [inline framework/src/base/EigenProblem.C]: I moved these  parameters to `Eigenvalue` (executioner)

File: framework/src/base/EigenProblem.C (hunk 21 above)
Comment (fdkong) [inline framework/src/base/EigenProblem.C]: Actually, I need `_nl_eigen` in `Eigenproblem`. But `FEProblemBase` needs `_nl_eigen`.

File: framework/src/base/EigenProblem.C (hunk 23 above)
Comment (fdkong) [inline framework/src/base/EigenProblem.C]: I should change `_nl` to `_nl_eigen`

File: framework/include/actions/DetermineSystemType.h (hunk 27 above)
Comment (permcody) [inline framework/include/actions/DetermineSystemType.h]: OK, then I might suggest that you move _action_params to just a local variable. You only need it during act() so there's no need to store it as an instance variable.

**Comment by fdkong** [comment]
@permcody

**Review by permcody — APPROVED** [review]

**Comment by brianmoose** [comment]
@fdkong You have a fix for relap-7? https://www.moosebuild.org/job/111936/

**Comment by fdkong** [comment]
@andrsd will look at this. Thanks.
